File: DAY 5/solutions.java/1011. Capacity To Ship Packages Within D Days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {

        int left = 0;
        int right = 0;

        for (int weight : weights) {
            left = Math.max(left, weight);
            right += weight;
        }

        while (left < right) {

            int capacity = left + (right - left) / 2;

            int currentWeight = 0;
            int requiredDays = 1;

            for (int weight : weights) {

                if (currentWeight + weight > capacity) {
                    requiredDays++;
                    currentWeight = 0;
                }

                currentWeight += weight;
            }

            if (requiredDays <= days) {
                right = capacity;
            } else {
                left = capacity + 1;
            }
        }

        return left;
    }
}

File: DAY 5/solutions.py/1011. Capacity To Ship Packages Within D Days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:

        left = max(weights)
        right = sum(weights)

        while left < right:

            capacity = (left + right) // 2

            currentWeight = 0
            requiredDays = 1

            for weight in weights:

                if currentWeight + weight > capacity:
                    requiredDays += 1
                    currentWeight = 0

                currentWeight += weight

            if requiredDays <= days:
                right = capacity
            else:
                left = capacity + 1

        return left